### ISSUE TITLE
Using Classify in Symbology dialog does not return all boolean values

### ISSUE THREAD

**Issue by DelazJ** [qgis/QGIS#40068]
In Layer Properties, Symbology, Categorized, using an expression returning a boolean does not seem to always return all possible values, ie `0`, `1` (and `all other values`)
See classes of `gid=attribute(@atlas_feature, 'gid')`, `gid=12` and `gid='toto'` below

note:: While the expression preview shows 0 or 1 in the screencast, I get the same "bug" for expressions that previews as true/false (eg, `is_selected( @atlas_feature)`)

![classifyButton](https://user-images.githubusercontent.com/7983394/99180567-90b63480-2727-11eb-8016-9d2ca84ea54c.gif)

QGIS version | 3.16.0-Hannover | QGIS code branch | Release 3.16 (9e55f5e99)
-- | -- | -- | --
Compiled against Qt | 5.12.8 | Running against Qt | 5.12.8
Compiled against GDAL/OGR | 3.0.4 | Running against GDAL/OGR | 3.0.4
Compiled against GEOS | 3.8.0-CAPI-1.13.1 | Running against GEOS | 3.8.0-CAPI-1.13.1
Compiled against SQLite | 3.31.1 | Running against SQLite | 3.31.1
PostgreSQL Client Version | 12.4 (Ubuntu 12.4-0ubuntu0.20.04.1) | SpatiaLite Version | 4.3.0a
QWT Version | 6.1.4 | QScintilla2 Version | 2.11.2
Compiled against PROJ | 6.3.1 | Running against PROJ | Rel. 6.3.1, February 10th, 2020
OS Version | Ubuntu 20.04.1 LTS | This copy of QGIS writes debugging output.
Active python plugins | db_manager; MetaSearch; processing

**Comment by roya0045** [comment]
if you use "gid"=12, does that work?

**Comment by roya0045** [comment]
I'm pretty sure this one isn't a bug, if you want to represent a value that isn't present just add it manually. The tool doesn't see it as boolean but as integer.

**Comment by DelazJ** [comment]
> The classify will only return observed values, not possible values. If the tool would return all possible values you would have millions of values all the time, this doesn't make sense.

`gid = 12` as an expression is a boolean check, not an integer. It checks whether the field has value 12 or not. not checking what possible values the field should have. And for a boolean, possible values are `true` or `false` nothing more. Now if you say; it returns observed values, it explains the `gid=12` outputs (`0`) but not the toto or atlas_feature ones (empty).

**Comment by roya0045** [comment]
As for toto the only value observed is false because gid is never = toto. the atlas feature one is a toss up depending if th atlas is active, not sure about the preview.

**Comment by DelazJ** [comment]
> As for toto the only value observed is false because gid is never = toto. 

Yes, the 'toto' case would also be OK, for observed values. I doesn't output an empty value as I said just above.

>the atlas feature one is a toss up depending if th atlas is active, not sure about the preview.

Of course the atlas is active, otherwise the preview wouldn't evaluate to 0 or 1 and these results are correct. If inactive, I guess you'd get either 0 or Null (didn't check and this is not the point here).
If "Classify" returns the observed values, it should return what the preview shows. If not, then one of them is buggy.

**Comment by roya0045** [comment]
Only the cas of the atlas feature is a valid bug to me. Evaluating an '=' can never be null, only 0 or 1.

**Comment by Pedro-Murteira** [comment]
Still valid on QGIS 3.22.4. and 3.24.0.
